src/big/big_int.cr

```diff
@@ -47,21 +47,40 @@ struct BigInt < Int
   end
 
   # Creates a `BigInt` from the given *num*.
-  def initialize(num : Int::Signed)
-    if LibC::Long::MIN <= num <= LibC::Long::MAX
-      LibGMP.init_set_si(out @mpz, num)
+  def self.new(num : Int::Primitive)
+    if LibGMP::SI::MIN <= num <= LibGMP::SI::MAX
+      LibGMP.init_set_si(out mpz1, LibGMP::SI.new!(num))
+      new(mpz1)
+    elsif LibGMP::UI::MIN <= num <= LibGMP::UI::MAX
+      LibGMP.init_set_ui(out mpz2, LibGMP::UI.new!(num))
+      new(mpz2)
```

Comment:
I'm wondering if it makes sense to reorder the conditions. Assuming that `BigInt` is typically used when operating with numbers that lie outside of UI/SI range, maybe the UI exclusion should come first to have the big numbers outside that range on a fast track? With the current order there are two range checks before reaching the else branch.
```cr
if LibGMP::UI::MIN <= num <= LibGMP::UI::MAX
  if LibGMP::SI::MIN <= num <= LibGMP::SI::MAX
    # init_set_si
  else
    # init_set_ui
  end
else
  # slow path
end
```

Reply:
`SI` and `UI`'s ranges do not overlap; the above code would always use the slow path for negative numbers.

Reply:
Ah, sorry. I messed up the conditions. This should work:
```cr
if LibGMP::SI::MIN <= num <= LibGMP::UI::MAX
  if num <= LibGMP::SI::MAX
    # init_set_si
  else
    # init_set_ui
  end
else
  # slow path
end
```